### ISSUE TITLE
[2.5.0] CSRF check is bypassed if request doesn't contain a cookie.

### ISSUE THREAD

**Issue by Manc** [playframework/playframework#5865]
I'm not 100% sure if I'm not missing something, but it doesn't feel right.

The CSRF check is bypassed if request doesn't contain at least one cookie.

### Play Version

2.5.0

### API (Scala / Java / Neither / Both)

Scala

### Operating System

MacOS 10.11

### JDK

```
java version "1.8.0_74"
Java(TM) SE Runtime Environment (build 1.8.0_74-b02)
Java HotSpot(TM) 64-Bit Server VM (build 25.74-b02, mixed mode)
```

### Expected Behavior

Please describe the expected behavior of the issue, starting from the first action.

1. Create an action with `CSRFCheck`.
2. Submit a `POST` request to that action without any cookies.
3. The `CSRFCheck` should filter request and respond a `403 Forbidden` and a message like `CSRF token check failed`.

### Actual Behavior

The request is not filtered and the action is executed as if it wasn't wrapped in a `CSRFCheck`.

The logs don't give much away:
```
--- (Running the application, auto-reloading is enabled) ---

[info] p.c.s.NettyServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000

(Server started, use Ctrl+D to stop and go back to the console...)

[info] application - ApplicationTimer demo: Starting application at 2016-03-11T17:51:25.657Z.
[debug] p.a.l.c.ActorSystemProvider - Starting application default Akka system: application
[debug] p.a.l.c.CryptoConfigParser - Generated dev mode secret 64f7e1c30aa20ad59be8cdbfcb5c3748 for app at file:/xxx/play-scala/target/scala-2.11/classes/application.conf
[info] play.api.Play - Application started (Dev)
[trace] p.a.mvc.Action - Invoking action with request: POST /form/check
[trace] p.c.s.n.PlayRequestHandler - channelReadComplete: ctx = io.netty.channel.DefaultChannelHandlerContext@35b2e1fb
```

### Reproducible Test Case

I've created a sample project at [https://github.com/Manc/play-scala-csrf](https://github.com/Manc/play-scala-csrf) to demonstrate and test the issue. All the instructions are in the README file.

**Comment by gmethvin** [comment]
@Manc This was a deliberate change, and is documented in the [migration guide](https://www.playframework.com/documentation/2.5.x/Migration25#CSRF-changes):

> There’s a new config option to bypass the new CSRF protection for requests with certain headers. This config option is turned on by default for the Cookie and Authorization headers, so that REST clients, which typically don’t use session authentication, will still work without having to send a CSRF token.

> However, since the config option allows through all requests without those headers, applications that use other authentication schemes (NTLM, TLS client certificates) will be vulnerable to CSRF. These applications should disable the config option so that their authenticated (cookieless) requests are protected by the CSRF filter.

You can restore the old behavior by setting the config option:
```
play.filters.csrf.protectHeaders = null
```

I'm also curious, what kind of authentication are you using?

**Comment by Manc** [comment]
@gmethvin Thank you very much for clarification.

It works (although you had a typo there; the correct config option would be `play.filters.csrf.header.protectHeaders = null`).

I guess I was confused by the bit _"applications that use other authentication schemes"_, because that doesn't cover unauthenticated requests such public forms (think of a contact form, or anonymous editing of Wiki pages, where the only identification is the user's IP address) and login forms.

Depending on the kind of services the web application offers, **login CSRF** can be a serious security risk, which may reach from user tracking to stealing a user's credit card details (see examples in section 3 in http://seclab.stanford.edu/websec/csrf/csrf.pdf).

Personally, I'm totally fine with the solution now, but I would suggest to reconsider the default behaviour. In my opinion developers should be encouraged to protect unauthenticated non-GET/HEAD requests too and disable this behaviour if they are sure that it's not required.

Thanks again for your help!

**Comment by gmethvin** [comment]
(Thanks for finding the typo. Updated.)

This was actually changed to fix the [Beacon vulnerability in Chrome](https://www.playframework.com/security/vulnerability/20160304-CsrfBypass), so now we check all content types by default.

Login CSRF is a legitimate concern, but the user has to do something else on the site for it to work, so its effectiveness depends a lot on how the site itself is designed. So we decided we would try to avoid breaking existing APIs and leave it to the site authors to decide.
